### PR TITLE
incr-join, find_updates: avoid unncecessary clones & use partition

### PR DESCRIPTION
**Pull request by Centril**
# Description of Changes

First commit: Only clone once per `TableOp` in `find_ops`, achieved using `Iterator::partition`:

```
Benchmarking incr-join: Collecting 100 samples in estimated 5.
incr-join               time:   [2.9260 µs 2.9296 µs 2.9334 µs]
                        change: [-5.7704% -5.0041% -4.4248%] (p = 0.00 < 0.05)
                        Performance has improved.
```

Second commit: Clone even less and have `JoinSide` hold `Vec<PV>` instead:
```
Benchmarking incr-join: Collecting 100 samples in estimated 5.
incr-join               time:   [2.5716 µs 2.5786 µs 2.5901 µs]
                        change: [-17.024% -16.337% -15.765%] (p = 0.00 < 0.05)
                        Performance has improved.
```

All numbers are vs. master baseline.
(i7-7700K, 64GB RAM)

# API and ABI breaking changes

None

# Expected complexity level and risk

1

# Testing

No semantic changes.

### CODE REVIEW

File: crates/core/src/subscription/subscription.rs [hunk 1]
```diff
@@ -360,41 +332,48 @@ impl IncrementalJoin {
...
+        let mut lhs_deletes = Vec::new();
+        let mut rhs_inserts = Vec::new();
+        let mut rhs_deletes = Vec::new();
+
+        // Partitions deletes of `update` into `ds` and inserts into `is`.
+        let partition_into = |ds: &mut Vec<_>, is: &mut Vec<_>, updates: &DatabaseTableUpdate| {
+            for update in &updates.ops {
+                if update.op_type == 0 { &mut *ds } else { &mut *is }.push(update.row.clone());
+            }
+        };
```
[the remark this answers is not in the document]
Comment (gefjon) [reply crates/core/src/subscription/subscription.rs:345]: Also, ugh, I hate all these `0`s and `1`s. If I get some spare time, I'll make a follow-up that defines constants `OP_TYPE_DELETE` and `OP_TYPE_INSERT` (assuming we don't have those already...).

File: crates/core/src/subscription/subscription.rs [hunk 2]
```diff
@@ -360,41 +332,48 @@ impl IncrementalJoin {
...
 
+        // Partitions all updates into the `l/rhs_insert/delete_ops` above.
```
Comment (gefjon) [inline crates/core/src/subscription/subscription.rs:347]:
```suggestion
        // Partitions all updates into the `(l|r)hs_(insert|delete)_ops` above.
```
I'm pretty sure this isn't a correct Perl-style regex, but I still find it more readable.

**Review by gefjon — APPROVED** [review]
Beautiful. The unnecessary clones in this code path have been bugging me for a while. Thanks for getting rid of them!

Minor comment-related crap, plus a question about reborrowing.

File: crates/core/src/subscription/subscription.rs [hunk 3]
```diff
@@ -360,46 +332,48 @@ impl IncrementalJoin {
...
     ) -> Option<(JoinSide, JoinSide)> {
-        let mut lhs_ops = Vec::new();
-        let mut rhs_ops = Vec::new();
+        let mut lhs_inserts = Vec::new();
+        let mut lhs_deletes = Vec::new();
+        let mut rhs_inserts = Vec::new();
+        let mut rhs_deletes = Vec::new();
+
+        // Partitions deletes of `update` into `ds` and inserts into `is`.
+        let partition_into = |ds: &mut Vec<_>, is: &mut Vec<_>, updates: &DatabaseTableUpdate| {
```
Comment (joshua-spacetime) [inline crates/core/src/subscription/subscription.rs:341]: Rename ds -> deletes, is -> inserts

**Review by joshua-spacetime — DISMISSED** [review]

**Pull request merged**